### ISSUE TITLE
Can’t connect to local server with port forwarding

### ISSUE THREAD

**Issue by rasnes** [rationalmatter/Juno-Issue-Tracker#60]
This is weird. I can’t connect to my own server after I did it with no issues just two days ago, after following your instructions to set a Jupyter server. It simply says “Authentication failed”. But I don’t understand that error message, as I can connect via ssh in other apps with the exact same user and password, at port 22. 

Can it be that the password field in Juno connect is unable to store the password, for some reason, as it seems to be blank when I try “fix” my settings? Or any ideas about some obvious things to check?

**Comment by navoshta** [comment]
Hi @rasnes, could you please attach a screenshot of the "Authentication failed" message — just to make sure that it's the SSH tunnel part that can't authenticate (and not Jupyter)?

> Can it be that the password field in Juno connect is unable to store the password, for some reason, as it seems to be blank when I try “fix” my settings?

Yes, this sounds like it could be related, as Juno Connect should show the concealed password when you try to edit your server configuration (if there is any password stored). Does it stay blank when you re-enter it and save the server configuration?

**Comment by rasnes** [comment]
The error message:

![image](https://user-images.githubusercontent.com/13234940/66345408-51c48580-e950-11e9-960b-9210afc6f022.png)

The Edit screen looks like this after an unsuccessful connection attempt:

![image](https://user-images.githubusercontent.com/13234940/66345145-b7644200-e94f-11e9-8248-dfe533b0820a.png)

The password area seems blank, as you can see, even though I enter it correctly. If I change the port from 22 to something else and try to connect (which obviously fails), it seems that the corresponding number of dots for the length of my password is present in the `Password` area, which was why I started suspecting that this field actually might go blank when I try to connect (on port 22).

**Comment by navoshta** [comment]
Thanks @rasnes — yes, judging by the screenshots, it's specifically SSH tunnel authentication that fails, and it does seem like the password is blank for some reason, hence authentication failure. What happens when you re-enter it and save the server configuration? Is it correctly saved if you create a _new_ server configuration with the same settings?

Either way, I'll look into this; although if you think of anything that could help me reproduce the issue, that would be great. Perhaps you updated iOS recently? Or maybe this configuration had a private key set previously, which you changed to password? Or (I understand it's a sensible question) maybe you use some odd characters in that password? Let me know if you can think of something that could trigger this.

Meanwhile, if adding a new configuration doesn't help, I'd try deleting and reinstalling the app.

**Comment by rasnes** [comment]
Thanks for looking into this! Reentering the password with the correct username behaves the exact same way as the first time I try to connect. But if enter a username that does not exist, it tries to authenticate and stores the password upon reopening the connect config. Creating a new server instance behaves exactly the same.

Nothing updated on the iPad. Just bought, installed the newest iPadOS, and then installed Juno and connected successfully via SSH tunneling, for now about ten days ago. But after a couple of days this issue started occurring. Always pwd authentication, have not tried with key. Only ASCII characters in password. Also tried reinstalling the app, but problem persisted.

Something that could have change was an update on the Ubuntu server, a package upgrade or something there. But I checked the firewall (and is able to connect via SHH with e.g. Terminux) and port 22 is open, so I can’t think of anything concrete creating the issue.

Now I set up a Direct connection to it, and that works well. Working on this server from outside of my local network is no pressing issue, so I am fine the Direct connection for time being 👍
